rfcs/20190117-tf-module.md

```diff
@@ -0,0 +1,634 @@
+    ...     # ^ Will be named "my_module/my_other_module"
+
+    >>> m = MyModule()
+    >>> # ^ Name scope will be closed thanks to metaclass.
+    """
+```
+
+### Naming
+
+Having human readable names for ops and variables is essential for debugging.
+Each `tf.Module` will have a name scope which is determined in the constructor.
+By default this will be a `lower_camel_case` version of the `ClassName`:
+
+```python
+class MyModule(tf.Module):
+  def __init__(self, name=None):
+    super(MyModule, self).__init__(name=name)
+    self.v = tf.Variable(1., name='v')
+
+>>> m = MyModule()
+>>> m.name
+'my_module'
+>>> m.name_scope.name
+'my_module/'
+>>> m.v.name
+'my_module/v:0'
+
+>>> m = MyModule(name='custom_name')
+>>> m.name
+'custom_name'
+>>> m.name_scope.name
+'custom_name/'
+>>> m.v.name
+'custom_name/v:0'
+```
+
+Modules created inside methods of other modules inherit their parent module
+name:
+
+```python
+class ParentModule(tf.Module):
+  def __init__(self):
+    super(ParentModule, self).__init__()
+    self.child_module = MyModule(name='child')
+    self.v = tf.Variable(1., name='v')
+
+>>> p = ParentModule()
+
+>>> p.name
+'parent_module'
+>>> p.name_scope.name
+'parent_module/'
+>>> p.v.name
+'parent_module/v:0'
+
+>>> p.child_module.name
+'child_module'
+>>> p.child_module.name_scope.name
+'parent_module/child/'
+>>> p.child_module.v.name
+'parent_module/child/v:0'
+```
+
+In TensorFlow 2 string matching of variable names is not required for
+distributed training (you can create remote variables eagerly, also see
+[DistributionStrategy](https://github.com/tensorflow/community/blob/master/rfcs/20181016-replicator.md))
+or checkpointing (this is done based on Python object structure) so the naming
+of variables and ops is just to aid with usability and visualisation.
+
+### Variables
+
+Users are free to create `tf.Variable` instances in any method at any time (as
+long as the super constructor has run).
+
+Variable reuse should be achieved using standard object oriented programming
+(e.g. assigning to object properties). `tf.Module` doesn't do anything behind the
+scenes to help (or hinder) variable reuse.
+
+We considered including a default implementation of `__call__` since lazy
+building is extremely common (e.g. constructing model parameters the first time
+an input is observed), however we backtracked on this for a number of reasons:
+
+-   Even among the authors there existed different preferences for how to enable
+    this pattern, it seemed unlikely we could make a choice which would satisfy
+    the community.
+-   Not all modules want a `__call__` method (e.g. an `Optimizer`) and we got
+    strong feedback that it is annoying to ignore unused methods inherited from
+    your base class.
+-   Encouraging a lazy pattern might make it harder to create modules with
+    multiple forward methods (e.g. a VAE module might be implemented with an
+    `encode` and `decode` method with no `__call__`).
```

Comment:
We prototyped it such that you didn't have to implement `__call__` (e.g. `_call` was not `@abstractmethod`), but if users called your module then we throw `NotImplementedError`. Even this has downsides though, like introspection/generated docs making it look like you had a call method.

Reply:
@tomhennigan, I agree 100% with this decision. There is no need to bloat the `tf.Module` class more than is absolutely required. Since the goal of this class is to handle stateful DAGs of modules, it is best to focus efforts on that.

As pointed out in this doc, there are many use cases where we would want to handle stateful DAGs but don't require a `__call__` method. I see it the responsibility of other libraries/sub-package to extend the functionality of `tf.Module`. I find specialized classes much easier to extend and reason about.